### PR TITLE
uart: Add UART implementation for Xilinx platform

### PR DESCRIPTION
**Pull request by cristiantpop**
This is used in tinyiiod project to communicate with iio-oscilloscope.
This version uses a fifo when receiving, data since there is no handshake
for smaller chunks of data.

Signed-off-by: Cristian Pop <cristian.pop@analog.com>

### CODE REVIEW

File: drivers/platform/xilinx/uart.c [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+		return FAILURE;
+	}
+
+	/*
+	 * Initialize the UART driver so that it's ready to use
+	 * Look up the configuration in the config table, then initialize it.
+	 */
+	config = XUartPs_LookupConfig(descriptor->device_id);
+	if (!config) {
+		return FAILURE;
```
Comment (mirceacaprioru) [inline drivers/platform/xilinx/uart.c]: free ? also in the next ones.

File: drivers/platform/xilinx/uart.c [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+	}
+
+	status = XUartPs_CfgInitialize(xil_uart_desc->instance, config,
+				       config->BaseAddress);
+	if (status != XST_SUCCESS) {
+		return FAILURE;
```
Comment (mirceacaprioru) [inline drivers/platform/xilinx/uart.c]: free?

File: drivers/platform/xilinx/uart.c [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <xparameters.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "error.h"
+#include "fifo.h"
```
Comment (adrimbarean) [inline drivers/platform/xilinx/uart.c]: Do we want the FIFO module dependency? Will we always use the UART module along with the FIFO, or we might want to use just the UART sometimes? I am perfectly OK if we will always use the UART with FIFO, just wanted it to be a conscious decision on our part, I guess.

File: drivers/platform/xilinx/uart.c [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+ * @param call_back_ref - Instance of UART.
+ * @param event - Event that caused the interrupt.
+ * @param data_len - number of bytes received.
+ */
+static void uart_irq_handler(void *call_back_ref, uint32_t event,
+			     uint32_t data_len)
+{
+	struct xil_uart_desc *xil_uart_desc = call_back_ref;
+	/* All of the data has been received */
+	if (event == XUARTPS_EVENT_RECV_DATA) {
```
Comment (adrimbarean) [inline drivers/platform/xilinx/uart.c]: Would this string of `if`'s be better as a `switch` clause?

**Comment by cristiantpop** [comment]
V3:

- fix memory leaks

File: drivers/platform/xilinx/uart.c (hunk 4 above)
Comment (cristiantpop) [reply drivers/platform/xilinx/uart.c]: Probably, I will change that.

File: drivers/platform/xilinx/uart.c (hunk 3 above)
Comment (cristiantpop) [reply drivers/platform/xilinx/uart.c]: We want as less dependencies as possible, but when receiving large amounts of data(from iio-oscilloscope), we are not able to process those fast enough and data gets lost. Introducing fifo fixes this.

File: drivers/platform/xilinx/uart.c [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+static int32_t uart_irq_init(struct uart_desc *descriptor)
+{
+	uint32_t uart_irq_mask;
+	struct xil_uart_desc *xil_uart_desc = descriptor->extra;
+
+	int32_t status = irq_register(xil_uart_desc->irq_desc, xil_uart_desc->irq_id,
+				      (Xil_ExceptionHandler) XUartPs_InterruptHandler, xil_uart_desc->instance);
+	if (status < 0)
+		return status;
+	XUartPs_SetHandler(xil_uart_desc->instance, uart_irq_handler, xil_uart_desc);
```
Comment (mirceacaprioru) [inline drivers/platform/xilinx/uart.c]: will this work with any xilinx platform ?

File: include/uart.h [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+*******************************************************************************/
+#ifndef UART_H_
+#define UART_H_
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+struct uart_init_par {
+	uint32_t	device_id;
```
Comment (mirceacaprioru) [inline include/uart.h]: is uint32 needed for device id?

**Comment by cristiantpop** [comment]
V4:

- Declare "uint8_t device_id".

- Add platform checks to see what platform it is used, and call platform specific functions. 

- Use "switch" statement instead of "if"s in "uart_irq_handler".

**Comment by cristiantpop** [comment]
V5:

- update commit message.

- update comment to: "/* Read data to UART. */" in uart.h

- change type from "uint32_t" to "uint8_t" for "device_id" in uart.h

File: drivers/platform/xilinx/uart.c [hunk 7]
```diff
@@ -0,0 +1,400 @@
+/******************************************************************************
+*
+* Copyright (C) 2010 - 2014 Xilinx, Inc.  All rights reserved.
```
Comment (mirceacaprioru) [inline drivers/platform/xilinx/uart.c]: i would include in the commit message that this patch offers support only for specific xilinx products, not the entire platform.

**Comment by cristiantpop** [comment]
V5:

- update commit message: specify supported platform.

**Comment by dbogdan** [comment]
Since include/uart.h is not Xilinx platform dependent, I would add it in a separate commit, that will maybe include a uart.c implementation for the generic platform.

**Comment by cristiantpop** [comment]
V6:

- Add a separate commit for "uart.h" generic header.

**Comment by cristiantpop** [comment]
V7:

- add UART generic implementation.

**Pull request merged**